diff --git a/nachos-project/code/machine/timer.cc b/nachos-project/code/machine/timer.cc
index c18c317..30e1f14 100644
--- a/nachos-project/code/machine/timer.cc
+++ b/nachos-project/code/machine/timer.cc
@@ -66,7 +66,6 @@ void Timer::CallBack() {
 void Timer::SetInterrupt() {
     if (!disable) {
         int delay = TimerTicks;
-
         if (randomize) {
             delay = 1 + (RandomNumber() % (TimerTicks * 2));
         }
diff --git a/nachos-project/code/test/Makefile b/nachos-project/code/test/Makefile
index a8c5ba1..10d7379 100644
--- a/nachos-project/code/test/Makefile
+++ b/nachos-project/code/test/Makefile
@@ -112,7 +112,7 @@ ifeq ($(hosttype),unknown)
 PROGRAMS = unknownhost
 else
 # change this if you create a new test program!
-PROGRAMS =worker ptest abs add halt shell matmult sort segments test_syscall num_io char_io random str_io ascii bubblesort help create_file open_file readwrite exec test_sem sinhvien voinuoc main
+PROGRAMS = sleeptest worker ptest abs add halt shell matmult sort segments test_syscall num_io char_io random str_io ascii bubblesort help create_file open_file readwrite exec test_sem sinhvien voinuoc main
 endif
 
 all: $(PROGRAMS)
@@ -126,6 +126,13 @@ halt: halt.o start.o
 	$(LD) $(LDFLAGS) start.o halt.o -o halt.coff
 	$(COFF2NOFF) halt.coff halt
 
+sleeptest.o: sleeptest.c
+	$(CC) $(CFLAGS) -c sleeptest.c
+
+sleeptest: sleeptest.o start.o
+	$(LD) $(LDFLAGS) start.o sleeptest.o -o sleeptest.coff
+	$(COFF2NOFF) sleeptest.coff sleeptest
+
 add.o: add.c
 	$(CC) $(CFLAGS) -c add.c
 
diff --git a/nachos-project/code/test/sleeptest.c b/nachos-project/code/test/sleeptest.c
new file mode 100644
index 0000000..0718c82
--- /dev/null
+++ b/nachos-project/code/test/sleeptest.c
@@ -0,0 +1,15 @@
+#include "syscall.h"
+int main(){
+	int p1 = ExecWithPriority("../test/worker",10); // parent process more priority(default - INT_MAX) 
+	int i;
+	for(i=0;i<10;i++){
+		PrintString("Parent before sleep\n");
+	}
+	PrintString("Parent Slept\n");
+	Sleep(10); // sleep(time in sec)- 10 sec -> converted to ticks  1 sec=10^6 ticks
+	for(i=0;i<10;i++){
+		PrintString("Parent after sleep\n");
+	}
+	Join(p1);
+	return 0;
+}
\ No newline at end of file
diff --git a/nachos-project/code/test/worker.c b/nachos-project/code/test/worker.c
index c11c182..9d2af53 100644
--- a/nachos-project/code/test/worker.c
+++ b/nachos-project/code/test/worker.c
@@ -3,10 +3,10 @@ int main() {
    int i, j,myPid;
    myPid=GetPid();
     for (i = 0; i < 10; i++) {
-        for (j = 0; j < 10000; j++); 
+        //for (j = 0; j < 10000; j++); 
     }
-    PrintString("Child worker process PID");
-    PrintNum(myPid);
-    PrintString(" (COMPLETED)\n");
+    //PrintString("Child worker process PID");
+    //PrintNum(myPid);
+    PrintString("child Completed\n");
     return 0;
 }
\ No newline at end of file
diff --git a/nachos-project/code/threads/alarm.cc b/nachos-project/code/threads/alarm.cc
index f15ba3b..e836f4f 100644
--- a/nachos-project/code/threads/alarm.cc
+++ b/nachos-project/code/threads/alarm.cc
@@ -19,8 +19,16 @@
 //      "doRandom" -- if true, arrange for the hardware interrupts to
 //		occur at random, instead of fixed, intervals.
 //----------------------------------------------------------------------
+int compr(Thread* t1,Thread* t2){
+     if(t1->wakeuptime < t2->wakeuptime) return -1;
+     else if(t1->wakeuptime > t2->wakeuptime) return 1;
+     else return 0;
+}
 
-Alarm::Alarm(bool doRandom) { timer = new Timer(doRandom, this); }
+Alarm::Alarm(bool doRandom) { 
+    timer = new Timer(doRandom, this);
+    sleepList=new SortedList<Thread*>(compr);
+}
 
 //----------------------------------------------------------------------
 // Alarm::CallBack
@@ -39,10 +47,15 @@ Alarm::Alarm(bool doRandom) { timer = new Timer(doRandom, this); }
 //	For now, just provide time-slicing.  Only need to time slice
 //      if we're currently running something (in other words, not idle).
 //----------------------------------------------------------------------
-
 void Alarm::CallBack() {
     Interrupt *interrupt = kernel->interrupt;
     MachineStatus status = interrupt->getStatus();
+    int now=kernel->stats->totalTicks;
+    while(!sleepList->IsEmpty() && sleepList->Front()->wakeuptime <= now){
+          Thread * t=sleepList->RemoveFront();
+          t->wakeuptime=0;
+          kernel->scheduler->ReadyToRun(t);    
+     }
     if (status != IdleMode) {
         if(kernel->scheduler->checkPreemption()){
             interrupt->YieldOnReturn();
diff --git a/nachos-project/code/threads/alarm.h b/nachos-project/code/threads/alarm.h
index af6dd02..d307904 100644
--- a/nachos-project/code/threads/alarm.h
+++ b/nachos-project/code/threads/alarm.h
@@ -21,10 +21,12 @@
 #include "utility.h"
 #include "callback.h"
 #include "timer.h"
-
+#include "list.h"
+#include "thread.h"
 // The following class defines a software alarm clock.
 class Alarm : public CallBackObj {
    public:
+    SortedList <Thread*>* sleepList; //keeps tracks of sleeping threads sorted by wakeuptime
     Alarm(bool doRandomYield);  // Initialize the timer, and callback
                                 // to "toCall" every time slice.
     ~Alarm() { delete timer; }
@@ -34,7 +36,6 @@ class Alarm : public CallBackObj {
 
    private:
     Timer *timer;  // the hardware timer device
-
     void CallBack();  // called when the hardware
                       // timer generates an interrupt
 };
diff --git a/nachos-project/code/threads/kernel.cc b/nachos-project/code/threads/kernel.cc
index 61f0bde..e77cb70 100644
--- a/nachos-project/code/threads/kernel.cc
+++ b/nachos-project/code/threads/kernel.cc
@@ -88,7 +88,7 @@ void Kernel::Initialize(char *userProgName /*=NULL*/) {
     // We didn't explicitly allocate the current thread we are running in.
     // But if it ever tries to give up the CPU, we better have a Thread
     // object to save its state.
-    currentThread = new Thread(userProgName,0); // giving first process a priority 0 - least 
+    currentThread = new Thread(userProgName); // giving first process a default priority (INT_MAX)- highest 
     currentThread->setStatus(RUNNING);
     stats = new Statistics();        // collect statistics
     interrupt = new Interrupt;       // start up interrupt handling
diff --git a/nachos-project/code/threads/thread.cc b/nachos-project/code/threads/thread.cc
index 58d35af..1352283 100644
--- a/nachos-project/code/threads/thread.cc
+++ b/nachos-project/code/threads/thread.cc
@@ -37,6 +37,7 @@ Thread::Thread(char *threadName,int prior, bool _has_dynamic_name /*=false*/) {
     has_dynamic_name = _has_dynamic_name;
     name = threadName;
     priority=prior; // give user given priority 
+    wakeuptime=0;
     stackTop = NULL;
     stack = NULL;
     status = JUST_CREATED;
@@ -231,20 +232,37 @@ void Thread::Yield() {
 //----------------------------------------------------------------------
 void Thread::Sleep(bool finishing) {
     Thread *nextThread;
-
     ASSERT(this == kernel->currentThread);
     ASSERT(kernel->interrupt->getLevel() == IntOff);
-
     DEBUG(dbgThread, "Sleeping thread: " << name);
-
     status = BLOCKED;
     while ((nextThread = kernel->scheduler->FindNextToRun()) == NULL)
         kernel->interrupt->Idle();  // no one to run, wait for an interrupt
-
     // returns when it's time for us to run
     kernel->scheduler->Run(nextThread, finishing);
 }
 
+void Thread::MySleep(int time) {
+    IntStatus oldstatus = kernel->interrupt->SetLevel(IntOff);
+    Thread *nextThread;
+    ASSERT(this == kernel->currentThread);
+    ASSERT(time >= 0);
+    ASSERT(kernel->interrupt->getLevel() == IntOff);
+    DEBUG(dbgThread, "Sleeping thread: " << name);
+    status = BLOCKED;
+    int timeticks_fromnow=time*1e6; // 1 sec= 10^6 ticks
+    ASSERT(kernel->stats->totalTicks + timeticks_fromnow < INT_MAX);
+    this->wakeuptime= kernel->stats->totalTicks + timeticks_fromnow;
+    kernel->alarm->sleepList->Insert(this);
+    kernel->interrupt->Schedule(kernel->alarm,timeticks_fromnow,TimerInt);
+    while ((nextThread = kernel->scheduler->FindNextToRun()) == NULL)
+        kernel->interrupt->Idle();  // no one to run, wait for an interrupt
+    // returns when it's time for us to run
+    kernel->scheduler->Run(nextThread,FALSE);
+    kernel->interrupt->SetLevel(oldstatus);
+}
+
+
 //----------------------------------------------------------------------
 // ThreadBegin, ThreadFinish,  ThreadPrint
 //	Dummy functions because C++ does not (easily) allow pointers to member
diff --git a/nachos-project/code/threads/thread.h b/nachos-project/code/threads/thread.h
index d11abec..d72dcf1 100644
--- a/nachos-project/code/threads/thread.h
+++ b/nachos-project/code/threads/thread.h
@@ -80,13 +80,14 @@ class Thread {
     bool has_dynamic_name;  // true if the thread name is dynamically allocated
 
    public:
-    Thread(char *debugName,int priority=0,
+    Thread(char *debugName,int priority=INT_MAX,
            bool _has_dynamic_name = false);  // initialize a Thread
     ~Thread();                               // deallocate a Thread
                                              // NOTE -- thread being deleted
                                              // must not be running when delete
                                              // is called
     int priority;
+    int wakeuptime; // for sleeping thread to wakeup 
     int processID;
     int parrentID;
     int exitStatus;
@@ -102,6 +103,8 @@ class Thread {
                                  // other thread is runnable
     void Sleep(bool finishing);  // Put the thread to sleep and
                                  // relinquish the processor
+
+    void MySleep(int time);
     void Begin();                // Startup code for the thread
     void Finish();               // The thread is done executing
 
diff --git a/nachos-project/code/userprog/exception.cc b/nachos-project/code/userprog/exception.cc
index bc25b52..530ac8c 100644
--- a/nachos-project/code/userprog/exception.cc
+++ b/nachos-project/code/userprog/exception.cc
@@ -162,6 +162,12 @@ void handle_SC_Abs(){
     return move_program_counter();
 }
 
+void handle_SC_Sleep(){
+        int time=(int)kernel->machine->ReadRegister(4);
+        SysSleep(time);
+        return move_program_counter();
+}
+
 void handle_SC_ReadNum() {
     int result = SysReadNum();
     kernel->machine->WriteRegister(2, result);
@@ -455,7 +461,9 @@ void ExceptionHandler(ExceptionType which) {
             return handle_SC_Abs();
          case SC_ExecWithPriority:
                     return handle_SC_ExecWithPriority();
-                case SC_ReadNum:
+          case SC_Sleep:
+              return handle_SC_Sleep();
+            case SC_ReadNum:
                     return handle_SC_ReadNum();
                 case SC_PrintNum:
                     return handle_SC_PrintNum();
diff --git a/nachos-project/code/userprog/ksyscall.h b/nachos-project/code/userprog/ksyscall.h
index 52aebc8..09647f6 100644
--- a/nachos-project/code/userprog/ksyscall.h
+++ b/nachos-project/code/userprog/ksyscall.h
@@ -220,6 +220,11 @@ int SysExecwithPriority(char* name,int priority) {
     return kernel->pTab->ExecUpdate(name,priority);
 }
 
+void SysSleep(int time){
+       kernel->currentThread->MySleep(time);
+}
+
+
 int SysJoin(int id) { return kernel->pTab->JoinUpdate(id); }
 
 int SysExit(int id) { return kernel->pTab->ExitUpdate(id); }
diff --git a/nachos-project/code/userprog/syscall.h b/nachos-project/code/userprog/syscall.h
index 67cd1b7..82711d5 100644
--- a/nachos-project/code/userprog/syscall.h
+++ b/nachos-project/code/userprog/syscall.h
@@ -49,6 +49,7 @@
 #define SC_GetPid 54
 #define SC_Abs 55
 #define SC_ExecWithPriority 56
+#define SC_Sleep 57
 #ifndef IN_ASM
 /* The system call interface.  These are the operations the Nachos
  * kernel needs to support, to be able to run user programs.
@@ -71,6 +72,8 @@ int Add(int op1, int op2);
 
 int Abs(int x);
 
+void Sleep(int t);
+
 int ReadNum();
 
 void PrintNum(int num);
